controlplane/pkg/nsm/nsm.go

```diff
@@ -290,78 +294,83 @@ func (srv *networkServiceManager) request(ctx context.Context, request networkse
-			logrus.Errorf("NSM:(9.2.1-%v) Dataplane request failed: %v retry: %v", requestID, err, dpRetry)
 
-			// Let's try again with a short delay
-			if dpRetry < DataplaneRetryCount-1 {
-				<-time.After(DataplaneRetryDelay)
-				continue
-			}
-			logrus.Errorf("NSM:(9.2.2-%v) Dataplane request  all retry attempts failed: %v", requestID, cc.Xcon)
-			// 9.3 If datplane configuration are failed, we need to close remore NSE actually.
-			srv.requestFailed(requestID, cc, existingCC, true, false)
-			return nil, err
+		logrus.Infof("NSM:(9.2.1-%v) Sending request to dataplane: %v retry: %v", requestID, editor.ClientConnection.Xcon, dpRetry)
+		newXcon, err = dataplaneClient.Request(dpCtx, editor.ClientConnection.Xcon)
+		if err == nil {
+			logrus.Infof("NSM:(9.2.2-%v) Dataplane configuration successful %v", requestID, editor.ClientConnection.Xcon)
+			break
 		}
 
-		// In case of context deadline, we need to close NSE and dataplane.
-		if err := ctx.Err(); err != nil {
-			srv.requestFailed(requestID, cc, existingCC, true, false)
-			return nil, ctx.Err()
+		logrus.Errorf("NSM:(9.2.3-%v) Dataplane request failed: %v retry: %v", requestID, err, dpRetry)
+
+		// Let's try again with a short delay
+		if dpRetry != DataplaneRetryCount-1 {
+			<-time.After(DataplaneRetryDelay)
 		}
+	}
 
-		logrus.Infof("NSM:(9.3-%v) Dataplane configuration successful %v", requestID, cc.Xcon)
-		break
+	if err != nil {
+		srv.requestFailed(requestID, editor, existingCC, true, false)
+		return nil, err
 	}
 
-	// 10. Send update for client connection
-	srv.model.ApplyClientConnectionChanges(cc.GetID(), func(cc *model.ClientConnection) {
-		cc.ConnectionState = model.ClientConnectionReady
-		cc.DataplaneState = model.DataplaneStateReady
-		cc.Xcon = newXcon
-	})
+	// 10. Update client connection
+	editor.ClientConnection.DataplaneState = model.DataplaneStateReady
+	editor.ClientConnection.Xcon = newXcon
+
+	if err = srv.model.CommitClientConnectionChanges(editor); err == nil {
+		_, err = srv.model.ChangeClientConnectionState(conn.GetId(), model.ClientConnectionReady)
+	}
+
+	if err != nil {
+		srv.requestFailed(requestID, editor, existingCC, true, true)
+		return nil, err
+	}
 
 	// 11. We are done with configuration here.
 	logrus.Infof("NSM:(11-%v) Request done...", requestID)
 
-	return cc.GetConnectionSource(), nil
+	return editor.ClientConnection.GetConnectionSource(), nil
 }
 
-func (srv *networkServiceManager) requestFailed(requestID string, cc, existingCC *model.ClientConnection, closeNSE, closeDp bool) {
+func (srv *networkServiceManager) requestFailed(requestID string, editor *model.ClientConnectionEditor, existingCC *model.ClientConnection, closeNSE, closeDp bool) {
 	logrus.Errorf("NSM:(%v) Request failed", requestID)
 
-	if cc == nil {
+	if editor == nil {
 		return
 	}
 
 	if closeNSE {
-		if err := srv.closeEndpoint(context.Background(), cc); err != nil {
+		if err := srv.closeEndpoint(context.Background(), editor.ClientConnection); err != nil {
```

Comment:
Probably accept `editor` instead of `clientConnection` is better. What do you think?